File: tests/oracle/block_findirreducible_1204.cc
/*
 * BLOCK-FINDIRREDUCIBLE-0001: locked Ghidra 12.0.4
 * BlockGraph::findIrreducible (block.cc:1147-1199) irreducible-edge
 * labeling oracle, including the structureLoops driver wiring
 * (block.cc:2194-2215).
 *
 * The fixture builds synthetic control-flow graphs through the production
 * BlockGraph APIs (newBlock/addEdge), runs findSpanningTree followed by
 * findIrreducible (or the full structureLoops driver), and observes the
 * complete post-call state: needrebuild and irreduciblecount results, the
 * preorder vector fed to findIrreducible, the final rootlist, the reordered
 * component list, every block's index / visitcount / numdesc / mark /
 * copymap (the FIND-structure collapse state), and every out-edge and
 * mirrored in-edge label.
 *
 * Cases cover: a reducible diamond with zero marks, a self-loop back edge
 * whose head is excluded from the reachunder set (block.cc:1160), the
 * classic two-entry irreducible pair with a forward edge promoted to
 * irreducible (block.cc:1182 cross/forward clear), parallel edges seeding
 * the reachunder set twice (double irreduciblecount), a nested
 * irreducible-inside-outer-loop graph whose outer reachunder walk observes
 * copymaps collapsed by the inner pass (FIND reuse, block.cc:1161/1173/
 * 1194), a multi-root graph where a cross edge is promoted to irreducible,
 * and the full structureLoops end-to-end driver on a reducible graph.
 */

#include <bits/stdc++.h>

// Test-only access: FlowBlock's private fields sit behind an explicit
// `private:` label, but BlockGraph's helper members (findSpanningTree,
// findIrreducible, ...) are implicitly private (no access label after
// `class BlockGraph : public FlowBlock {`), so `private -> public` alone
// cannot reach them.  `class -> struct` opens the implicit-private section;
// no decompile header uses `template<class ...>` or `enum class`, and
// <bits/stdc++.h> has already pulled every standard header, so the rewrite
// only touches Ghidra declarations in this translation unit.
#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

using namespace ghidra;

static std::string flagsText(uint4 label)
{
  std::string out;
  if (label & FlowBlock::f_goto_edge) out += 'g';
  if (label & FlowBlock::f_loop_edge) out += 'l';
  if (label & FlowBlock::f_defaultswitch_edge) out += 'd';
  if (label & FlowBlock::f_irreducible) out += 'i';
  if (label & FlowBlock::f_tree_edge) out += 't';
  if (label & FlowBlock::f_forward_edge) out += 'f';
  if (label & FlowBlock::f_cross_edge) out += 'c';
  if (label & FlowBlock::f_back_edge) out += 'b';
  if (label & FlowBlock::f_loop_exit_edge) out += 'x';
  if (out.empty()) out.push_back('-');
  return out;
}

class Graph {
public:
  BlockGraph graph;
  std::vector<FlowBlock *> created;

  FlowBlock *makeBlock(void)
  {
    FlowBlock *bl = graph.newBlock();
    created.push_back(bl);
    return bl;
  }

  void edge(FlowBlock *from, FlowBlock *to)
  {
    graph.addEdge(from, to);
  }

  std::string name(FlowBlock *bl) const
  {
    std::ostringstream out;
    out << 'b' << (std::find(created.begin(), created.end(), bl) - created.begin());
    return out.str();
  }

  // Uniform post-call projection shared by both call paths. `rebuildText`
  // and `cntText` are "-" for the structureLoops case (not observable from
  // outside the driver; the local irreduciblecount never escapes).
  void observe(const std::string &caseName,
               const std::string &rebuildText,
               const std::string &cntText,
               const std::vector<FlowBlock *> &preorder,
               const std::vector<FlowBlock *> &rootlist)
  {
    std::ostringstream pre, roots, list, blocks, outedges, inedges;
    for (int4 i = 0; i < preorder.size(); ++i) {
      if (i != 0) pre << ',';
      pre << name(preorder[i]);
    }
    for (int4 i = 0; i < rootlist.size(); ++i) {
      if (i != 0) roots << ',';
      roots << name(rootlist[i]);
    }
    const std::vector<FlowBlock *> &gl = graph.getList();
    for (int4 i = 0; i < gl.size(); ++i) {
      if (i != 0) list << ',';
      list << name(gl[i]);
    }
    blocks << '[';
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      if (i != 0) blocks << ';';
      blocks << name(bl) << ":index=" << bl->index
             << ",visit=" << bl->visitcount
             << ",desc=" << bl->numdesc
             << ",mark=" << ((bl->flags & FlowBlock::f_mark) ? 1 : 0)
             << ",copy=";
      if (bl->copymap == bl)
        blocks << "self";
      else if (bl->copymap == (FlowBlock *)0)
        blocks << "null";
      else
        blocks << "other:" << name(bl->copymap);
    }
    blocks << ']';
    bool first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->outofthis.size(); ++s) {
        if (!first) outedges << ';';
        first = false;
        outedges << name(bl) << '!' << s << '>' << name(bl->outofthis[s].point)
                 << ':' << flagsText(bl->outofthis[s].label);
      }
    }
    first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->intothis.size(); ++s) {
        if (!first) inedges << ';';
        first = false;
        inedges << name(bl) << '@' << s << '<' << name(bl->intothis[s].point)
                << ':' << flagsText(bl->intothis[s].label);
      }
    }
    std::cout << "case=" << caseName
              << "|rebuild=" << rebuildText
              << "|cnt=" << cntText
              << "|pre=[" << pre.str() << ']'
              << "|roots=[" << roots.str() << ']'
              << "|list=[" << list.str() << ']'
              << "|blocks=" << blocks.str()
              << "|out=[" << outedges.str() << ']'
              << "|in=[" << inedges.str() << ']' << '\n';
  }

  // findSpanningTree + findIrreducible path (block.cc:2203-2204 shape).
  void runPair(const std::string &caseName)
  {
    std::vector<FlowBlock *> preorder;
    std::vector<FlowBlock *> rootlist;
    graph.findSpanningTree(preorder, rootlist);
    int4 irreduciblecount = 0;
    bool needrebuild = graph.findIrreducible(preorder, irreduciblecount);
    std::ostringstream rb, cn;
    rb << (needrebuild ? 1 : 0);
    cn << irreduciblecount;
    observe(caseName, rb.str(), cn.str(), preorder, rootlist);
  }
};

int main(void)
{
  std::cout << std::unitbuf;
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);
  std::cout << "schema=1|fixture=BLOCK-FINDIRREDUCIBLE-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    {
      // Reducible single-entry diamond plus bypass: no back edges, the
      // reachunder set stays empty for every vertex, zero marks.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b0, b3);
      g.runPair("reducible_diamond_zero_marks");
    }
    {
      // Self-loop: the back edge b1->b1 has y == x, so the loop head never
      // enters its own reachunder set (block.cc:1160); zero marks.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b1);
      g.edge(b1, b2);
      g.runPair("self_loop_head_skipped");
    }
    {
      // Classic two-entry irreducible pair: b1 <-> b2 with entries from b0
      // on both sides. The forward edge b0!1->b2 is promoted to irreducible
      // (its forward classification is cleared, block.cc:1182) and b2
      // collapses into the loop head b1.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b2);
      g.edge(b2, b1);
      g.runPair("classic_two_entry");
    }
    {
      // Parallel edges b0->b0->b2: the reachunder seeding loop pushes
      // FIND(y) once per back edge without a dedup check (block.cc:1161),
      // so the BFS re-scans the duplicate member and irreduciblecount
      // accumulates per parallel edge (2).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b0, b2);
      g.edge(b1, b2);
      g.edge(b2, b1);
      g.runPair("parallel_edges_double_count");
    }
    {
      // Irreducible pair {b2,b3} nested inside the outer b1 loop: x=b2's
      // pass collapses b3 into b2, and the later x=b1 pass observes
      // FIND(b3) = b2 (copymap reuse, block.cc:1161/1173/1194); the whole
      // inner component collapses into b1. b1!1->b3 is promoted.
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      FlowBlock *b4 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b3, b2);
      g.edge(b2, b4);
      g.edge(b4, b1);
      g.runPair("nested_irreducible_find_reuse");
    }
    {
      // Two roots (b0, b1): the rootlist head/tail swap makes b1's tree the
      // first visited; b0's later cross edge b0!0->b2 is promoted to
      // irreducible (cross classification cleared).
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b3, b2);
      g.runPair("multi_root_cross_to_irreducible");
    }
    {
      // Full structureLoops driver (block.cc:2194-2215) on a reducible
      // graph: single pass, no rebuild, irreduciblecount == 0 so calcLoop
      // is never reached. needrebuild/irreduciblecount are driver-internal
      // (projected as "-").
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      FlowBlock *b4 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b2);
      g.edge(b2, b1);
      g.edge(b2, b3);
      g.edge(b3, b2);
      g.edge(b3, b4);
      std::vector<FlowBlock *> rootlist;
      g.graph.structureLoops(rootlist);
      std::vector<FlowBlock *> preorder;
      g.observe("structure_loops_reducible_endtoend", "-", "-", preorder, rootlist);
    }
  }
  catch (const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/block_findirreducible_1204.rs
// BLOCK-FINDIRREDUCIBLE-0001: Rugra comparand for the locked Ghidra 12.0.4
// BlockGraph::findIrreducible (block.cc:1147-1199) oracle.  Mirrors
// tests/oracle/block_findirreducible_1204.cc case for case: the same
// synthetic control-flow graphs are built through the production BlockGraph
// APIs (add_block/add_edge), find_spanning_tree + find_irreducible (or the
// full structure_loops driver) run, and the complete post-call state is
// printed in the shared observation format.

use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::block::{edge_flags as ef, BlockBasic, BlockGraph, FlowBlock};

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

fn flags_text(label: u32) -> String {
    let mut out = String::new();
    if label & ef::F_GOTO_EDGE != 0 { out.push('g'); }
    if label & ef::F_LOOP_EDGE != 0 { out.push('l'); }
    // Rugra's F_DEFAULTSWITCH_EDGE shares bit 7 with F_TREE_EDGE
    // (pre-existing edge_flags collision). Within this projection bit 7 can
    // only be a tree label: findSpanningTree wipes every edge label at the
    // start of each pass and neither pass ever sets default-switch labels.
    if label & ef::F_DEFAULTSWITCH_EDGE != 0 && label & ef::F_TREE_EDGE == 0 { out.push('d'); }
    if label & ef::F_IRREDUCIBLE_EDGE != 0 { out.push('i'); }
    if label & ef::F_TREE_EDGE != 0 { out.push('t'); }
    if label & ef::F_FORWARD_EDGE != 0 { out.push('f'); }
    if label & ef::F_CROSS_EDGE != 0 { out.push('c'); }
    if label & ef::F_BACK_EDGE != 0 { out.push('b'); }
    if label & ef::F_LOOP_EXIT_EDGE != 0 { out.push('x'); }
    if out.is_empty() { out.push('-'); }
    out
}

struct Graph {
    graph: BlockGraph,
    created: Vec<BlockRef>,
    base: u64,
}

impl Graph {
    fn new(base: u64) -> Self {
        Graph { graph: BlockGraph::new(), created: Vec::new(), base }
    }

    fn make_block(&mut self) -> BlockRef {
        let bl: BlockRef = Arc::new(RwLock::new(BlockBasic::new(
            0,
            Address::new(self.base),
        )));
        self.graph.add_block(bl.clone());
        self.created.push(bl.clone());
        bl
    }

    fn edge(&mut self, from: &BlockRef, to: &BlockRef) {
        self.graph.add_edge(from.clone(), to.clone());
    }

    fn name(&self, bl: &BlockRef) -> String {
        for (i, cand) in self.created.iter().enumerate() {
            if Arc::ptr_eq(cand, bl) {
                return format!("b{}", i);
            }
        }
        "b?".to_string()
    }

    // Uniform post-call projection shared by both call paths. `rebuild_text`
    // and `cnt_text` are "-" for the structure_loops case (not observable
    // from outside the driver).
    fn observe(
        &self,
        case_name: &str,
        rebuild_text: &str,
        cnt_text: &str,
        preorder: &[BlockRef],
        rootlist: &[BlockRef],
    ) {
        let pre: Vec<String> = preorder.iter().map(|b| self.name(b)).collect();
        let roots: Vec<String> = rootlist.iter().map(|b| self.name(b)).collect();
        let list: Vec<String> = self.graph.blocks.iter().map(|b| self.name(b)).collect();
        let mut blocks = String::from("[");
        for (i, bl) in self.created.iter().enumerate() {
            if i != 0 { blocks.push(';'); }
            let g = bl.read().unwrap();
            blocks.push_str(&format!(
                "{}:index={},visit={},desc={},mark={},copy=",
                self.name(bl),
                g.get_index(),
                g.get_visit_count(),
                g.get_num_desc(),
                if g.is_mark() { 1 } else { 0 },
            ));
            let copy = match g.get_copy_map() {
                Some(w) => w,
                None => {
                    blocks.push_str("null");
                    continue;
                }
            };
            drop(g);
            match copy.upgrade() {
                Some(arc) if Arc::ptr_eq(&arc, bl) => blocks.push_str("self"),
                Some(arc) => blocks.push_str(&format!("other:{}", self.name(&arc))),
                None => blocks.push_str("null"),
            }
        }
        blocks.push(']');
        let mut outedges = String::from("[");
        let mut first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_out() {
                if let Some(e) = g.get_out(s) {
                    if !first { outedges.push(';'); }
                    first = false;
                    outedges.push_str(&format!(
                        "{}!{}>{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        outedges.push(']');
        let mut inedges = String::from("[");
        first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_in() {
                if let Some(e) = g.get_in(s) {
                    if !first { inedges.push(';'); }
                    first = false;
                    inedges.push_str(&format!(
                        "{}@{}<{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        inedges.push(']');
        println!(
            "case={}|rebuild={}|cnt={}|pre=[{}]|roots=[{}]|list=[{}]|blocks={}|out={}|in={}",
            case_name,
            rebuild_text,
            cnt_text,
            pre.join(","),
            roots.join(","),
            list.join(","),
            blocks,
            outedges,
            inedges,
        );
    }

    // find_spanning_tree + find_irreducible path (block.cc:2203-2204 shape).
    fn run_pair(&mut self, case_name: &str) {
        let mut preorder: Vec<BlockRef> = Vec::new();
        let mut rootlist: Vec<BlockRef> = Vec::new();
        self.graph
            .find_spanning_tree(&mut preorder, &mut rootlist)
            .expect("spanning tree");
        let mut irreduciblecount: i32 = 0;
        let needrebuild = self.graph.find_irreducible(&preorder, &mut irreduciblecount);
        self.observe(
            case_name,
            if needrebuild { "1" } else { "0" },
            &irreduciblecount.to_string(),
            &preorder,
            &rootlist,
        );
    }
}

fn main() {
    println!("schema=1|fixture=BLOCK-FINDIRREDUCIBLE-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // case=reducible_diamond_zero_marks
    {
        let mut g = Graph::new(0x5f00);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b0, &b3);
        g.run_pair("reducible_diamond_zero_marks");
    }

    // case=self_loop_head_skipped
    {
        let mut g = Graph::new(0x5f80);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b1);
        g.edge(&b1, &b2);
        g.run_pair("self_loop_head_skipped");
    }

    // case=classic_two_entry
    {
        let mut g = Graph::new(0x6000);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.run_pair("classic_two_entry");
    }

    // case=parallel_edges_double_count
    {
        let mut g = Graph::new(0x6080);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b0, &b2);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.run_pair("parallel_edges_double_count");
    }

    // case=nested_irreducible_find_reuse
    {
        let mut g = Graph::new(0x6100);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b4 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        g.edge(&b2, &b4);
        g.edge(&b4, &b1);
        g.run_pair("nested_irreducible_find_reuse");
    }

    // case=multi_root_cross_to_irreducible
    {
        let mut g = Graph::new(0x6180);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        g.run_pair("multi_root_cross_to_irreducible");
    }

    // case=structure_loops_reducible_endtoend
    {
        let mut g = Graph::new(0x6200);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b4 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        g.edge(&b3, &b4);
        let mut rootlist: Vec<BlockRef> = Vec::new();
        g.graph.structure_loops(&mut rootlist).expect("structure loops");
        let preorder: Vec<BlockRef> = Vec::new();
        g.observe("structure_loops_reducible_endtoend", "-", "-", &preorder, &rootlist);
    }
}
